Firestore/core/include/firebase/firestore/timestamp.h

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_INCLUDE_FIREBASE_FIRESTORE_TIMESTAMP_H_
+#define FIRESTORE_CORE_INCLUDE_FIREBASE_FIRESTORE_TIMESTAMP_H_
+
+#if !defined(_STLPORT_VERSION)
+#include <chrono>
+#endif
+#include <cstdint>
+#include <ctime>
+
+namespace firebase {
+
+/**
+ * A Timestamp represents a point in time independent of any time zone or
+ * calendar, represented as seconds and fractions of seconds at nanosecond
+ * resolution in UTC Epoch time. It is encoded using the Proleptic Gregorian
+ * Calendar which extends the Gregorian calendar backwards to year one. It is
+ * encoded assuming all minutes are 60 seconds long, i.e. leap seconds are
+ * "smeared" so that no leap second table is needed for interpretation. Range is
+ * from 0001-01-01T00:00:00Z to 9999-12-31T23:59:59.999999999Z.
+ *
+ * @see
+ * https://github.com/google/protobuf/blob/master/src/google/protobuf/timestamp.proto
+ */
+class Timestamp {
```

Comment:
Note: I had to cut down the interface I originally proposed:
* having `Timestamp` output a human-readable string to `ostream` is nice, but it forces dependency on `<ostream>` on users for very marginal gain. Since it's a public type, it seems like a bad trade-off;
* `std::tm` depends on the time zone, I'd rather leave it to users to correctly convert it to `time_t` if necessary;
* `ToTimeT()` seems more trouble than it's worth. It's not clear what to do if the number of seconds stored in `Timestamp` exceeds the limits of `time_t`, it's unspecified (I think) whether `time_t` supports negative numbers, and also it will at best amount to `{ return seconds(); }`;
* `ToTimePoint()` has a problem with truncation. `<chrono>` will try to represent the time as a single number, so the natural choice of precision in this case will be nanoseconds. However, the range of seconds supported by `Timestamp` overflows guaranteed limits of `long long`. This could be solved e.g. by truncating to microseconds, but this seems pretty non-intuitive. It's possible to provide something along the lines of `std::chrono::time_point<...> secondsAsTimePoint() const` and `std::chrono::time_point<...> nanosecondsAsTimePoint() const`, though, but the gain for users is probably too small.

Reply:
Some notes:
* A cut down set of forward declarations for iostream are available in `<iosfwd>`, which mostly addresses the concern, though realistically all we really need to do is something reasonable, and `std::string ToString()` would work just as well from my point of view. Since we're providing this for debugging it doesn't need to be fast.
* `std::tm` requires calendar information as well to calculate the day of the week. This definitely seems like something to leave to the user.
* `time_t` is largely unspecified, though practically it's just a `long` and essentially equivalent to seconds except on 32-bit platforms. I'm fine with omitting it though I think it's straightforward to include it if time_t is large enough.
* Timepoint requires template parameters, so if we wanted to support this API we could do so by allowing the caller to specify a Clock and Duration. At that point we'd be delegating to the user to avoid undefined behavior by requiring them to select values they care about. Note that if we selected system clock and milliseconds as our default everything mostly just works (though at lower resolutions than people might expect).